packages/core/src/input/pointer/PointerManager.ts

```diff
@@ -0,0 +1,402 @@
+          }
+          if (activePointerCount === 1) {
+            _effectiveEventList[this._effectiveEventCount++] = PointerEventType.Down;
+            /** Update pointer's threshold. */
+            this._minPointerThreshold = evt.pointerId;
+          }
+          break;
+        case "pointerup":
+          if (pointerIndex >= 0) {
+            this._updatePointer(pointerIndex, evt.offsetX, evt.offsetY, PointerPhase.Up, timeStamp);
+            activePointerCount === 1 && (_effectiveEventList[this._effectiveEventCount++] = PointerEventType.Up);
+          }
+          break;
+        case "pointermove":
+          if (_lastMoveHash[evt.pointerId - _minPointerThreshold] === i) {
+            if (pointerIndex < 0) {
+              this._addPointer(evt.pointerId, evt.type, evt.offsetX, evt.offsetY, PointerPhase.Move, timeStamp);
+              ++activePointerCount;
+            } else {
+              this._updatePointer(pointerIndex, evt.offsetX, evt.offsetY, PointerPhase.Move, timeStamp);
+            }
+            _effectiveEventList[this._effectiveEventCount++] = PointerEventType.Move;
+          }
+          break;
+        case "pointerout":
+          if (pointerIndex >= 0) {
+            this._removePointer(pointerIndex, timeStamp);
+            --activePointerCount === 0 && (_effectiveEventList[this._effectiveEventCount++] = PointerEventType.Leave);
+            this._curFrameHasCancel = true;
+          }
+          break;
+        default:
+          break;
+      }
+    }
+    /** Reset event list. */
+    eventList.length = 0;
+  }
+
+  /**
+   * Get the Entity to which the ray is cast.
+   * @param x - The X coordinate of the pointer on the screen, specified in normalized
+   * @param y - The Y coordinate of the pointer on the screen, specified in normalized
+   * @returns The Entity to which the ray is cast
+   */
+  private _pointerRaycast(): Entity {
+    let { x, y } = this._curFramePosition;
+    /** Convert screen coordinates to viewport coordinates. */
+    x /= this._canvas.width;
+    y /= this._canvas.height;
+    const cameras = this._engine.sceneManager.activeScene._activeCameras;
+    for (let i = cameras.length - 1; i >= 0; i--) {
+      const camera = cameras[i];
+      if (!camera.enabled || camera.renderTarget) {
+        continue;
+      }
+      const { x: vpX, y: vpY, z: vpW, w: vpH } = camera.viewport;
+      if (x >= vpX && y >= vpY && x - vpX <= vpW && y - vpY <= vpH) {
+        PointerManager._tempPoint.setValue((x - vpX) / vpW, (y - vpY) / vpH);
+        // TODO: Only check which colliders have listened to the input.
+        return this._engine.physicsManager.raycast(
+          camera.viewportPointToRay(PointerManager._tempPoint, PointerManager._tempRay),
+          PointerManager._tempHitResult
+        )
+          ? PointerManager._tempHitResult.collider.entity
+          : null;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Execute drag event.
+   */
+  private _exeDrag(): void {
```

Comment:
Try not to use shorthand for API，Very unclear and need to write extra notes. Write '_ExecuteDrag()'